driver/src/orderbook/batched_auction_data_reader.rs

```diff
@@ -0,0 +1,317 @@
+            },
+            index,
+            user_order_counts: HashMap::new(),
+        }
+    }
+
+    /// The pagination data used when reading data from the smart contract in
+    /// batches.
+    pub fn pagination(&self) -> &Pagination {
+        &self.pagination
+    }
+
+    /// Signal that no more data will be read and return the result consuming
+    /// self.
+    pub fn get_auction_data(self) -> (AccountState, Vec<Order>) {
+        (self.account_state, self.orders)
+    }
+
+    /// Applies one batch of data.
+    ///
+    /// A batch can come from `getEncodedUsersPaginated` or `getEncodedOrders`.
+    /// In the latter case there is only one batch.
+    ///
+    /// Returns the number of orders in the data.
+    ///
+    /// Panics if length of `packed_auction_bytes` is not a multiple of
+    /// `AUCTION_ELEMENT_WIDTH`.
+    pub fn apply_batch(&mut self, packed_auction_bytes: &[u8]) -> usize {
+        let previous_order_count = self.orders.len();
+        self.apply_auction_data(&packed_auction_bytes);
+        let number_of_added_orders = self.orders.len() - previous_order_count;
+        if number_of_added_orders == 0 {
+            return 0;
+        }
+        let last_order_user = self.orders.last().expect("there are no orders").account_id;
+        self.pagination.previous_page_user = last_order_user;
+        self.pagination.previous_page_user_offset = *self
+            .user_order_counts
+            .get(&last_order_user)
+            .expect("user has order but no order count");
+        number_of_added_orders
+    }
+
+    fn apply_auction_data(&mut self, packed_auction_bytes: &[u8]) {
+        let auction_elements = self.parse_auction_elements(packed_auction_bytes);
+        self.apply_auction_elements_to_account_state(auction_elements.iter());
+        self.orders.extend(
+            auction_elements
+                .into_iter()
+                .map(|auction_element| auction_element.order),
+        );
+    }
+
+    fn apply_auction_elements_to_account_state<'a, Iter>(&mut self, auction_elements: Iter)
+    where
+        Iter: Iterator<Item = &'a StableXAuctionElement>,
+    {
+        for element in auction_elements.into_iter() {
+            self.account_state.modify_balance(
+                element.order.account_id,
+                element.order.sell_token,
+                |x| {
+                    if *x == 0 {
+                        *x = element.sell_token_balance
+                    } else {
+                        assert_eq!(
+                            *x,
+                            element.sell_token_balance,
+                            "got order which sets user {}'s sell token {} \
+                            balance to {} but sell_token_balance has already \
+                            been set to {}",
+                            element.order.account_id,
+                            element.order.sell_token,
+                            element.sell_token_balance,
```

Comment:
What is this and why are we modifying account balance? Only the smart contract should be modifying account balances.

Reply:
Its modifying account balances in the `AccountState` struct in order to be able return it once the order reading is done. We have been doing the same thing since before this PR, the code just moved to a different file.
I don't know why we need `AccountState` in the grand scheme things, if that is what you're asking.